File: src/main/java/com/insogroup/utils/IOC/strategies/CacheIOCStrategy.java
package com.insogroup.utils.IOC.strategies;

import com.insogroup.utils.ClassExtractor.exceptions.ClassExtractException;
import com.insogroup.utils.ClassExtractor.ClassExtractor;
import com.insogroup.utils.IOC.exceptions.ResolveStrategyException;
import com.insogroup.utils.IOC.interfaces.IOCStrategy;

import java.util.HashMap;
import java.util.Map;

public class CacheIOCStrategy implements IOCStrategy {
    protected String targetClassPath;

    protected Map<String, Object> instances = new HashMap<>();

    /**
    * @param targetClassPath package.name for using with @{@link ClassExtractor}
    */
    public CacheIOCStrategy(String targetClassPath) throws ClassExtractException {
        this.targetClassPath = targetClassPath;
    }

    /**
     * Return cached instance of object
     * @param args not used
     * @return Instance that was generated in constructor
     * @throws ResolveStrategyException Throw when in constructor was not created object
     */
    @Override
    public Object getInstance(Object... args) throws ResolveStrategyException {
        if (args.length == 0) {
            return getInstance("");
        }
        String name = (String) args[0];
        if (instances.containsKey(name)) {
            return instances.get(name);
        }
        Object[] constructorArgs = new Object[args.length-1];//first argument is classpath
        System.arraycopy(args, 1, constructorArgs, 0, constructorArgs.length);
        try {
            Object instance = ClassExtractor.extract(targetClassPath, constructorArgs);
            instances.put(name, instance);
            return instance;
        } catch (ClassExtractException e) {
            throw new ResolveStrategyException("Can't find variable for this name and create new instance", e);
        }
    }
}

File: src/main/kotlin/com/github/insanusmokrassar/utils/IOC/strategies/CacheIOCStrategy.kt
package com.github.insanusmokrassar.utils.IOC.strategies

import com.github.insanusmokrassar.iobjectk.interfaces.IObject
import com.github.insanusmokrassar.utils.*
import com.github.insanusmokrassar.utils.ClassExtractor.exceptions.ClassExtractException
import com.github.insanusmokrassar.utils.IOC.IOC
import com.github.insanusmokrassar.utils.IOC.exceptions.ResolveStrategyException
import com.github.insanusmokrassar.utils.IOC.getOrCreateIOCInstance
import com.github.insanusmokrassar.utils.IOC.interfaces.IOCStrategy

import java.util.HashMap

class CacheIOCStrategy(config: IObject<Any>) : IOCStrategy {

    protected var instances: MutableMap<String, Any> = HashMap()
    protected val ioc: IOC = getOrCreateIOCInstance(config.get(IOCNameField))

    /**
     * Await:
     * <pre>
     *     {
     *          "IOCName": "ioc name",
     *          "preset": [
     *              {
     *                  "name": "NAME OF FIRST PRESET",
     *                  "package": "CLASS PATH",
     *                  "args": [
     *                      ARGS,
     *                      OF,
     *                      FIRST,
     *                      PRESET
     *                  ]
     *              }
     *          ]
     *     }
     * </pre>
     */
    init {
        if (config.keys().contains(presetField)) {
            val presetObject = config.get<List<IObject<Any>>>(presetField)
            presetObject.forEach {
                val args: Array<Any>
                if (it.keys().contains(argsField)) {
                    args = it.get<List<Any>>(argsField).toTypedArray()
                } else {
                    args = Array(0, {})
                }
                getInstance(
                        it.get<String>(nameField),
                        it.get<String>(packageField),
                        *args
                )
//                getInstance(it, *currentPresetArgs.toTypedArray())
            }
        }
    }

    /**
     * Return cached instance of object
     * @param args args[0] - name of dependency, args[1] - package or IOC dependency name, args[2..size] - args for dependency init
     *
     * @return Instance that was generated in constructor
     * @throws ResolveStrategyException Throw when in constructor was not created object
     */
    @Throws(ResolveStrategyException::class)
    override fun <T> getInstance(vararg args: Any): T {
        if (args.isEmpty()) {
            return getInstance("")
        }
        val name = args[0] as String
        if (instances.containsKey(name)) {
            return instances[name]!! as T
        }
        val targetClassPath = args[1] as String
        val constructorArgs = args.sliceArray(2..args.size-1)
        try {
            val instance = ioc.resolve<T>(targetClassPath, *constructorArgs)
            instances.put(name, instance!!)
            return instance
        } catch (e: ClassExtractException) {
            throw ResolveStrategyException("Can't find variable for this name and create new instance", e)
        }

    }
}
